## LOAN DEFAULT PREDICTION

#### Here the data will be cleaned and merged

In [58]:
# Import pandas and numpy

import pandas as pd
import numpy as np

In [59]:
#load all the datasets

# 1. Demographics dataset

demo = pd.read_csv('datasets/traindemographics.csv')
demo.head()

customerid                   birthdate  \
0  8a858e135cb22031015cbafc76964ebd  1973-10-10 00:00:00.000000   
1  8a858e275c7ea5ec015c82482d7c3996  1986-01-21 00:00:00.000000   
2  8a858e5b5bd99460015bdc95cd485634  1987-04-01 00:00:00.000000   
3  8a858efd5ca70688015cabd1f1e94b55  1991-07-19 00:00:00.000000   
4  8a858e785acd3412015acd48f4920d04  1982-11-22 00:00:00.000000   

  bank_account_type  longitude_gps  latitude_gps bank_name_clients  \
0           Savings       3.319219      6.528604           GT Bank   
1           Savings       3.325598      7.119403     Sterling Bank   
2           Savings       5.746100      5.563174     Fidelity Bank   
3           Savings       3.362850      6.642485           GT Bank   
4           Savings       8.455332     11.971410           GT Bank   

  bank_branch_clients employment_status_clients level_of_education_clients  
0                 NaN                       NaN                        NaN  
1                 NaN                 Permanent                        NaN  
2                 NaN                       NaN                        NaN  
3                 NaN                 Permanent                        NaN  
4                 NaN                 Permanent                        NaN

In [60]:
demo.shape

(4346, 9)

Demographics dataset has 4,364 rows and 9 columns

In [61]:
# 2. Previous loans dataset

prev_loan = pd.read_csv('/Users/RyanMburu/Desktop/DS Projects/Loan Default Prediction/datasets/trainprevloans.csv')
prev_loan.head()

customerid  systemloanid  loannumber  \
0  8a2a81a74ce8c05d014cfb32a0da1049     301682320           2   
1  8a2a81a74ce8c05d014cfb32a0da1049     301883808           9   
2  8a2a81a74ce8c05d014cfb32a0da1049     301831714           8   
3  8a8588f35438fe12015444567666018e     301861541           5   
4  8a85890754145ace015429211b513e16     301941754           2   

                 approveddate                creationdate  loanamount  \
0  2016-08-15 18:22:40.000000  2016-08-15 17:22:32.000000     10000.0   
1  2017-04-28 18:39:07.000000  2017-04-28 17:38:53.000000     10000.0   
2  2017-03-05 10:56:25.000000  2017-03-05 09:56:19.000000     20000.0   
3  2017-04-09 18:25:55.000000  2017-04-09 17:25:42.000000     10000.0   
4  2017-06-17 09:29:57.000000  2017-06-17 08:29:50.000000     10000.0   

   totaldue  termdays                  closeddate referredby  \
0   13000.0        30  2016-09-01 16:06:48.000000        NaN   
1   13000.0        30  2017-05-28 14:44:49.000000        NaN   
2   23800.0        30  2017-04-26 22:18:56.000000        NaN   
3   11500.0        15  2017-04-24 01:35:52.000000        NaN   
4   11500.0        15  2017-07-14 21:18:43.000000        NaN   

                 firstduedate             firstrepaiddate  
0  2016-09-14 00:00:00.000000  2016-09-01 15:51:43.000000  
1  2017-05-30 00:00:00.000000  2017-05-26 00:00:00.000000  
2  2017-04-04 00:00:00.000000  2017-04-26 22:03:47.000000  
3  2017-04-24 00:00:00.000000  2017-04-24 00:48:43.000000  
4  2017-07-03 00:00:00.000000  2017-07-14 21:08:35.000000

In [62]:
prev_loan.shape

(18183, 12)

The loans dataset has 18,183 rows with 12 columns. This is so as a single costumer can have more than one loan

In [63]:
# 3. Perfomance dataset

perf = pd.read_csv('datasets/trainperf.csv')
perf.head()

customerid  systemloanid  loannumber  \
0  8a2a81a74ce8c05d014cfb32a0da1049     301994762          12   
1  8a85886e54beabf90154c0a29ae757c0     301965204           2   
2  8a8588f35438fe12015444567666018e     301966580           7   
3  8a85890754145ace015429211b513e16     301999343           3   
4  8a858970548359cc0154883481981866     301962360           9   

                 approveddate                creationdate  loanamount  \
0  2017-07-25 08:22:56.000000  2017-07-25 07:22:47.000000     30000.0   
1  2017-07-05 17:04:41.000000  2017-07-05 16:04:18.000000     15000.0   
2  2017-07-06 14:52:57.000000  2017-07-06 13:52:51.000000     20000.0   
3  2017-07-27 19:00:41.000000  2017-07-27 18:00:35.000000     10000.0   
4  2017-07-03 23:42:45.000000  2017-07-03 22:42:39.000000     40000.0   

   totaldue  termdays referredby good_bad_flag  
0   34500.0        30        NaN          Good  
1   17250.0        30        NaN          Good  
2   22250.0        15        NaN          Good  
3   11500.0        15        NaN          Good  
4   44000.0        30        NaN          Good

In [64]:
perf.shape

(4368, 10)

Will iteratively clean all 3 datasets before merging.

The data should be : 
1. Valid
2. Accurate
3. Complete
4. Consistent
5. Uniform

## 1. Validity

Columns not useful in our analysis will be dropped 

In [65]:
# 1. Demographics Dataset

# will drop branch bank clients and level of education

demo.drop(columns=['bank_branch_clients', 'level_of_education_clients'], inplace=True, axis=1)

demo.head()

customerid                   birthdate  \
0  8a858e135cb22031015cbafc76964ebd  1973-10-10 00:00:00.000000   
1  8a858e275c7ea5ec015c82482d7c3996  1986-01-21 00:00:00.000000   
2  8a858e5b5bd99460015bdc95cd485634  1987-04-01 00:00:00.000000   
3  8a858efd5ca70688015cabd1f1e94b55  1991-07-19 00:00:00.000000   
4  8a858e785acd3412015acd48f4920d04  1982-11-22 00:00:00.000000   

  bank_account_type  longitude_gps  latitude_gps bank_name_clients  \
0           Savings       3.319219      6.528604           GT Bank   
1           Savings       3.325598      7.119403     Sterling Bank   
2           Savings       5.746100      5.563174     Fidelity Bank   
3           Savings       3.362850      6.642485           GT Bank   
4           Savings       8.455332     11.971410           GT Bank   

  employment_status_clients  
0                       NaN  
1                 Permanent  
2                       NaN  
3                 Permanent  
4                 Permanent

The other datasets are deemed valid.

## 2.  Accuracy

As its a business based datasets based on loans, it will be assumed the math is correct enough.

## 3. Completeness

Will check for null values and handle them if present

In [66]:
#1. Demographics dataset

demo.isna().sum()

customerid                     0
birthdate                      0
bank_account_type              0
longitude_gps                  0
latitude_gps                   0
bank_name_clients              0
employment_status_clients    648
dtype: int64

In [67]:
# so as to avoid losing data, will replace the NaN values in employment status column with unknown

demo['employment_status_clients'].fillna('Unknown', inplace = True)

In [68]:
#Check if the null values are replaced
demo.isna().sum()

customerid                   0
birthdate                    0
bank_account_type            0
longitude_gps                0
latitude_gps                 0
bank_name_clients            0
employment_status_clients    0
dtype: int64

In [69]:
# 2. Previouls loan dataset

prev_loan.isna().sum()

customerid             0
systemloanid           0
loannumber             0
approveddate           0
creationdate           0
loanamount             0
totaldue               0
termdays               0
closeddate             0
referredby         17157
firstduedate           0
firstrepaiddate        0
dtype: int64

The refered by column has 17,000 + null values, thats morev than 90% of the data. Dropping them would lead to too much data being lost.

The column itself will be dropped.


In [70]:
#Dropping the columns with alot of null values
prev_loan.drop(columns = ['referredby'], axis = 1, inplace = True)

In [71]:
# Check for null values again
prev_loan.isna().sum()

customerid         0
systemloanid       0
loannumber         0
approveddate       0
creationdate       0
loanamount         0
totaldue           0
termdays           0
closeddate         0
firstduedate       0
firstrepaiddate    0
dtype: int64

In [72]:
# 3. Performance dataset
perf.isna().sum()

customerid          0
systemloanid        0
loannumber          0
approveddate        0
creationdate        0
loanamount          0
totaldue            0
termdays            0
referredby       3781
good_bad_flag       0
dtype: int64

In [73]:
# will do the same as the loans dataset, dropping the referred by column
perf.drop(columns = ['referredby'], axis = 1, inplace = True)

In [74]:
perf.isna().sum()

customerid       0
systemloanid     0
loannumber       0
approveddate     0
creationdate     0
loanamount       0
totaldue         0
termdays         0
good_bad_flag    0
dtype: int64

## 4. Consistency

The data has to be consistent throughout. Here is where duplicates are often checked and handled

In [75]:
# Function for checking duplicates

def dups(dataset):
    total_dups = dataset.duplicated().sum()
    return total_dups


In [76]:
# Demographics dataset
dups(demo)

12

In [77]:
# drop those 12 duplicated values
demo = demo.drop_duplicates()

In [78]:
dups(demo)

0

In [79]:
# performance dataset
dups(perf)

0

In [80]:
# prev loans dataset
dups(prev_loan)

0

No more duplicates in our data

## 5. Uniformity

the dataset should look uniform throughout. Column names should be uniform and same for datatypes

In [81]:
# The column names look good, all in lower case and spaced out by underscores

# will proceed to check datatypes of the tables

def datatypes(dataframe):
    out = dataframe.dtypes
    return out

In [82]:
#demographics table
datatypes(demo)

customerid                    object
birthdate                     object
bank_account_type             object
longitude_gps                float64
latitude_gps                 float64
bank_name_clients             object
employment_status_clients     object
dtype: object

In [83]:
# Performance table
datatypes(perf)

customerid        object
systemloanid       int64
loannumber         int64
approveddate      object
creationdate      object
loanamount       float64
totaldue         float64
termdays           int64
good_bad_flag     object
dtype: object

In [84]:
# Previous loans table
datatypes(prev_loan)

customerid          object
systemloanid         int64
loannumber           int64
approveddate        object
creationdate        object
loanamount         float64
totaldue           float64
termdays             int64
closeddate          object
firstduedate        object
firstrepaiddate     object
dtype: object

Datatypes look good so far.

- Will proceed to calculate the ages of the customers based on their DOB in the demographics dataset

In [85]:
demo.head()

customerid                   birthdate  \
0  8a858e135cb22031015cbafc76964ebd  1973-10-10 00:00:00.000000   
1  8a858e275c7ea5ec015c82482d7c3996  1986-01-21 00:00:00.000000   
2  8a858e5b5bd99460015bdc95cd485634  1987-04-01 00:00:00.000000   
3  8a858efd5ca70688015cabd1f1e94b55  1991-07-19 00:00:00.000000   
4  8a858e785acd3412015acd48f4920d04  1982-11-22 00:00:00.000000   

  bank_account_type  longitude_gps  latitude_gps bank_name_clients  \
0           Savings       3.319219      6.528604           GT Bank   
1           Savings       3.325598      7.119403     Sterling Bank   
2           Savings       5.746100      5.563174     Fidelity Bank   
3           Savings       3.362850      6.642485           GT Bank   
4           Savings       8.455332     11.971410           GT Bank   

  employment_status_clients  
0                   Unknown  
1                 Permanent  
2                   Unknown  
3                 Permanent  
4                 Permanent

### will calculate the ages of customers as the final stage of cleaning

In [86]:
# Will attempt to calculate age from the customers' birthdates

demo['birthdate'] = demo['birthdate'].str.strip('00:00:00.000000').astype(str)

In [94]:
demo.head()

customerid    birthdate bank_account_type  \
0  8a858e135cb22031015cbafc76964ebd  1973-10-10            Savings   
1  8a858e275c7ea5ec015c82482d7c3996  1986-01-21            Savings   
2  8a858e5b5bd99460015bdc95cd485634  1987-04-01            Savings   
3  8a858efd5ca70688015cabd1f1e94b55  1991-07-19            Savings   
4  8a858e785acd3412015acd48f4920d04  1982-11-22            Savings   

   longitude_gps  latitude_gps bank_name_clients employment_status_clients  
0       3.319219      6.528604           GT Bank                   Unknown  
1       3.325598      7.119403     Sterling Bank                 Permanent  
2       5.746100      5.563174     Fidelity Bank                   Unknown  
3       3.362850      6.642485           GT Bank                 Permanent  
4       8.455332     11.971410           GT Bank                 Permanent

In [96]:
demo['birthdate'] = pd.to_datetime(demo['birthdate'])

In [97]:
demo.dtypes

customerid                           object
birthdate                    datetime64[ns]
bank_account_type                    object
longitude_gps                       float64
latitude_gps                        float64
bank_name_clients                    object
employment_status_clients            object
dtype: object

In [100]:
# Calculate age

# Will use strptime to change DOB to correct format

now = pd.to_datetime('now')
now

demo['age'] = (now - demo['birthdate']).astype('<m8[Y]')
demo.head()

customerid  birthdate bank_account_type  \
0  8a858e135cb22031015cbafc76964ebd 1973-10-10           Savings   
1  8a858e275c7ea5ec015c82482d7c3996 1986-01-21           Savings   
2  8a858e5b5bd99460015bdc95cd485634 1987-04-01           Savings   
3  8a858efd5ca70688015cabd1f1e94b55 1991-07-19           Savings   
4  8a858e785acd3412015acd48f4920d04 1982-11-22           Savings   

   longitude_gps  latitude_gps bank_name_clients employment_status_clients  \
0       3.319219      6.528604           GT Bank                   Unknown   
1       3.325598      7.119403     Sterling Bank                 Permanent   
2       5.746100      5.563174     Fidelity Bank                   Unknown   
3       3.362850      6.642485           GT Bank                 Permanent   
4       8.455332     11.971410           GT Bank                 Permanent   

    age  
0  48.0  
1  36.0  
2  34.0  
3  30.0  
4  39.0

In [101]:
perf.head()

customerid  systemloanid  loannumber  \
0  8a2a81a74ce8c05d014cfb32a0da1049     301994762          12   
1  8a85886e54beabf90154c0a29ae757c0     301965204           2   
2  8a8588f35438fe12015444567666018e     301966580           7   
3  8a85890754145ace015429211b513e16     301999343           3   
4  8a858970548359cc0154883481981866     301962360           9   

                 approveddate                creationdate  loanamount  \
0  2017-07-25 08:22:56.000000  2017-07-25 07:22:47.000000     30000.0   
1  2017-07-05 17:04:41.000000  2017-07-05 16:04:18.000000     15000.0   
2  2017-07-06 14:52:57.000000  2017-07-06 13:52:51.000000     20000.0   
3  2017-07-27 19:00:41.000000  2017-07-27 18:00:35.000000     10000.0   
4  2017-07-03 23:42:45.000000  2017-07-03 22:42:39.000000     40000.0   

   totaldue  termdays good_bad_flag  
0   34500.0        30          Good  
1   17250.0        30          Good  
2   22250.0        15          Good  
3   11500.0        15          Good  
4   44000.0        30          Good

In [102]:
prev_loan.head()

customerid  systemloanid  loannumber  \
0  8a2a81a74ce8c05d014cfb32a0da1049     301682320           2   
1  8a2a81a74ce8c05d014cfb32a0da1049     301883808           9   
2  8a2a81a74ce8c05d014cfb32a0da1049     301831714           8   
3  8a8588f35438fe12015444567666018e     301861541           5   
4  8a85890754145ace015429211b513e16     301941754           2   

                 approveddate                creationdate  loanamount  \
0  2016-08-15 18:22:40.000000  2016-08-15 17:22:32.000000     10000.0   
1  2017-04-28 18:39:07.000000  2017-04-28 17:38:53.000000     10000.0   
2  2017-03-05 10:56:25.000000  2017-03-05 09:56:19.000000     20000.0   
3  2017-04-09 18:25:55.000000  2017-04-09 17:25:42.000000     10000.0   
4  2017-06-17 09:29:57.000000  2017-06-17 08:29:50.000000     10000.0   

   totaldue  termdays                  closeddate                firstduedate  \
0   13000.0        30  2016-09-01 16:06:48.000000  2016-09-14 00:00:00.000000   
1   13000.0        30  2017-05-28 14:44:49.000000  2017-05-30 00:00:00.000000   
2   23800.0        30  2017-04-26 22:18:56.000000  2017-04-04 00:00:00.000000   
3   11500.0        15  2017-04-24 01:35:52.000000  2017-04-24 00:00:00.000000   
4   11500.0        15  2017-07-14 21:18:43.000000  2017-07-03 00:00:00.000000   

              firstrepaiddate  
0  2016-09-01 15:51:43.000000  
1  2017-05-26 00:00:00.000000  
2  2017-04-26 22:03:47.000000  
3  2017-04-24 00:48:43.000000  
4  2017-07-14 21:08:35.000000

#### Will go ahead and export the clean datasets 

In [105]:
# function that exports the datasets

def data_exporter(dataset):
    dataset_name = input()
    dataset.to_csv(dataset_name, index = False)

In [106]:
# Demographics data
data_exporter(demo)

In [108]:
# Performance data
data_exporter(perf)

In [109]:
# Previous loans data
data_exporter(prev_loan)